### ISSUE TITLE
Provide CloudEvent design proposal to extend KEDA

### ISSUE THREAD

**Issue by tomkerkhove** [kedacore/keda#479]
Provide CloudEvent that are emitted by the runtime so that customers can extend KEDA.

Events that come to mind are:
- `New Trigger Authentication Created`
- `Deployment Scaled to Zero`
- `New Scaled Object Created`
- Information on scaling action
- Information about new scaled object referencing existing trigger authentication

Only concern I have is that folks could go through the Kubernetes eventing approach to get the information, but if we would emit it ourselves we could make it more user friendly rather than having to figure this out via the Kube way

## Tasklist

- [x] Create proposal
- [x] Agree on proposal
- [x] Define list of events & payloads to emit
- [ ] Implement
- [ ] Document

**Comment by tomkerkhove** [comment]
Thoughts @jeffhollan?

**Comment by zroubalik** [comment]
We should definitely implement the Kuberenetes eventing approach (created https://github.com/kedacore/keda/issues/530). This could be lower priority item.

**Comment by tomkerkhove** [comment]
Thanks for that! Since Kubernetes doens't support CloudEvents (yet), I would leverage this as a seperate component as well to make it easier to integrate with other products/tools that are not Kubernetes event specific.

See https://github.com/kubernetes/kubernetes/issues/85544 for CloudEvents support in Kubernetes.

**Comment by tomkerkhove** [comment]
The reason for that is that, for most scenarios you'd want to react on these events, and these workloads are not always in Kubernetes. If we only support Kubernetes events how it is today, events are stuck inside the cluster so they are, in my opinion, often pointless.

If we have CloudEvents, we can just forward them to things like Azure Event Grid and take it from there.

**Comment by tomkerkhove** [comment]
I will provide a more thorough proposal before we start on this.

**Comment by tomkerkhove** [comment]
My goal with this issue is to allow end-users to subscribe to events emitted by KEDA that provide them insights into what is going on.

These should serve several purposes such as:
- Allow for autoscaling awareness (scaling)
- Allow for automated notifications for incidents (errors, misconfiguration)
- ...

> **But, but, Kubernetes events?!**
> Today, we already offer cluster events but the use-case for these are different since they live in the cluster and are hard to use outside of it unless you use something such as [Kubernetes Event Exporter](https://github.com/opsgenie/kubernetes-event-exporter) and parse the raw events which is clunky given there is no standard way of writing those events

From a high-level perspective, KEDA will be doing 3 things:

1. Scale workloads in a single namespace across one or more namespaces
2. Provide Kubernetes events inside the cluster that can be used through the typical tooling (CLI, UIs, ...)
3. Emit events to endpoints inside/outside the cluster that are [CloudEvents](http://cloudevents.io/)-compliant

![Picture1](https://user-images.githubusercontent.com/4345663/145172657-0500068f-490b-4ba2-b2d7-dfbb74e88248.png)

## Proposal

My first idea was that we could easily add this as an endpoint to which KEDA can push events, but this will quickly become a bottleneck for end-users that have numerous namespaces and want to have more control on who is allowed to receive what events:

- Cluster admins/operators will want to have events for scaling in all namespaces
- App devs/operators will only be interested in the events for their applications

While every team *could* do the filtering on their own, this is a waste of compute.

I believe it would be ideal if we introduce a new `EventSink` CRD (name is just proposal) which is scoped to the whole cluster (admin operators) or namespaces (app operators):

```yaml
apiVersion: keda.sh/v1alpha1
kind: EventSink # Or ClusterEventSink
metadata:
  name: operations-cross-cluster-events
spec:
  # Support regular webhook endpoints over HTTP
  webhook:
    uri: http://foo.bar
    authentication:
      apiKey:
        headerName: x-api-key
        valueFrom:
            secretKeyRef:
             name: secrets-operations-events
             key: webhook-api-key
  # We can add integration with other systems more easily, following TriggerAuthentication
  # This can be cloud providers, platforms, products, ...
  # ...
```

KEDA will then emit events to every `EventSink`, where appropriate.

If we should use our operator or another container for this to remove the burden/load off of our operator is something we can discuss.

**Comment by zroubalik** [comment]
I like this proposal, one minor nit, I'd use `apivVersion: events.keda.sh/v1alpha1` or similar, to distinguish from standard (scaling related) CRDs.

**Comment by tomkerkhove** [comment]
Great idea, updated!

**Comment by tomkerkhove** [comment]
As part of this or after the first iteration we should also provide:
- A "health" indication in our CRD to see if it was able to connect or not
- Prometheus metrics showing the success rate and amount of events forwarded.

**Comment by tomkerkhove** [comment]
A new event that would be nice to have is "we have reached max instance count, but we believe we should keep on scaling" to help customer revise their max instance count.

**Comment by tomkerkhove** [comment]
I've completed my design proposal and anyone who is interested is welcome to review and comment on it: https://docs.google.com/document/d/14Z04AWiMBkSX88uiaVS-kg-8QjjNcDhVy6-u5FMUk_w/edit?usp=sharing

In terms of implementation, I would identify a first batch of events to support and create dedicated items per event scenario so that they can be added afterwards in a phased-approach.

/cc @kedacore/keda-maintainers

**Comment by tomkerkhove** [comment]
Any thoughts @kedacore/keda-maintainers?

**Comment by tomkerkhove** [comment]
Any thoughts https://github.com/orgs/kedacore/teams/keda-maintainers?

**Comment by JorTurFer** [comment]
I have been checking it and in general I agree with all. You have done great job!!!!
The only point I don't agree is with the event when they have reached the max replicas. IMO we shouldn't do it because that means monitoring the workload (at least I feel that).

It's a really good feature for an hipotetycal KEDA v3 😉😉

**Comment by zroubalik** [comment]
I don't want v3 ever 😄 

Sorry for the delay on this, it is still on my radar, I just need to find a time for a proper review.

**Comment by JorTurFer** [comment]
I see the business value and I think it's easy to do xD The thing is that it requires information about the infrastructure status and I know KEDA can get it, but I'm not sure if KEDA should do it. I mean, why only max replicas and not min/current replicas? 
I feel that monitoring this information like infrastructure/workload monitoring and that should be done by other tools, prometheus/newrelic/datadog/whatever.

It's not a blocker for me, if we think that it's needed, we can do it

**Comment by tomkerkhove** [comment]
I don't see this as monitoring since neither Prometheus nor the rest will tell you when the workload needs 11 replicas but it's not scaling because 10 is the max.

Imagine having an event for which you can subscribe, page somebody and mitigate it before your system is flooded.

**Comment by tomkerkhove** [comment]
But similarly to how we offer scale-to-zero, we can leverage this as well in my opinion. If it's technically possible, which I think it should be but I do agree that this could be in a separate component if need be.

**Comment by dmeytin** [comment]
Can you please help me to understand how rising Cloud Events will help with multi-cluster use-case #1587 ?

**Comment by tomkerkhove** [comment]
Events will allow you to extend KEDA to close our current multi-cluster gap and take required actions. For example if we're at max replicas you can scale out in another region or if it fails to scale, etc.

However, we are open to event requests that can help you more though.

**Comment by tomkerkhove** [comment]
> I've completed my design proposal and anyone who is interested is welcome to review and comment on it: [docs.google.com/document/d/14Z04AWiMBkSX88uiaVS-kg-8QjjNcDhVy6-u5FMUk_w/edit?usp=sharing](https://docs.google.com/document/d/14Z04AWiMBkSX88uiaVS-kg-8QjjNcDhVy6-u5FMUk_w/edit?usp=sharing)
> 
> In terms of implementation, I would identify a first batch of events to support and create dedicated items per event scenario so that they can be added afterwards in a phased-approach.
> 
> /cc @kedacore/keda-maintainers

Last call for reviews before we close it on Tuesday EOB European time.

/cc @kedacore/keda-maintainers

**Comment by JorTurFer** [comment]
I have taken another look and I agree with all but `Maximum replicas exceeded for workload`, I still think that it's not under KEDA scope because KEDA doesn't know the current needed replicas, it's the HPA Controller who knows that.
You have done a great job designing this

**Comment by tomkerkhove** [comment]
> I have taken another look and I agree with all but Maximum replicas exceeded for workload, I still think that it's not under KEDA scope because KEDA doesn't know the current needed replicas, it's the HPA Controller who knows that.

I tend to disagree and say that this is implementation detail because KEDA's focus is making application autoscaling simple and using HPA is just an implementation detail that some/most KEDA end-users don't even know about.

Because of that, it's up to us to provide a unified autoscaling solution imo.

**Comment by dmeytin** [comment]
I hope we'll agree to start mult-cluster support for KEDA.
It's a crucial feature for enterprise scale systems.

**Comment by tomkerkhove** [comment]
@dmeytin Feel free to post in https://github.com/kedacore/keda/issues/1587

**Comment by tomkerkhove** [comment]
All work items have been created and can be progress can be found on https://github.com/orgs/kedacore/projects/2/views/23?query=is%3Aopen+sort%3Aupdated-desc
